internal/app/wwctl/configure/nfs/main.go

```diff
@@ -56,35 +48,57 @@ func Configure(show bool) error {
 		fmt.Fprintf(fstab, "sysfs /sys sysfs defaults 0 0\n")
 		fmt.Fprintf(fstab, "proc /proc proc defaults 0 0\n")
 
-		for _, export := range controller.Nfs.Exports {
-			fmt.Fprintf(exports, "%s %s/%s(sync)\n", export, controller.Network, controller.Netmask)
-			fmt.Fprintf(fstab, "%s:%s %s nfs defaults 0 0\n", controller.Ipaddr, export, export)
-		}
-
-		fmt.Printf("Enabling and restarting the NFS services\n")
-		if controller.Nfs.SystemdName == "" {
-			err := util.SystemdStart("nfs-server")
+		if controller.Nfs.Enabled {
```

Comment:
what happens when then key "Nfs.Enabled" does not exist?
are there unit tests for any of this logic?

Reply:
I think we probably want `Nfs.Enabled` to default to true in the case that it is not specified in the config file.

I'm not familiar enough with the code to understand how/where to properly test this. Could you provide some pointers? I've also made the PR editable by the maintainers in case it's more desirable to push a test than to explain how to implement it.

Note that I don't think the logic was tested before, or we would have caught the case where setting `Nfs.Enabled` to false had no effect.

Reply:
Unit tests??? ;-)

Reply:
I'm testing this code now and I got `line 25: cannot unmarshal !!map into []*warewulfconf.NfsExportConf` error with config bellow. How should I write a config??

```
 19 nfs:
 20   systemd name: nfs-server
 21   exports:
 22   - /home
 23   - /var/warewulf
 24   exports extended:
 25     path: /opt
 26     options: rw
 27     mount: true
``

Reply:
Sorry, my bad. The following works:

```
 19 nfs:
 20   enabled: true
 21   systemd name: nfs-server
 22   exports:
 23   - /home
 24   - /var/warewulf
 25   exports extended:
 26     - path: /opt
 27       options:
 28         - rw
 29         - sync
 30       mount: true
```

Reply:
Sorry, my bad. Above error disappear with the following config.
but now I have error in `/etc/exports`.
Options(string array) are convert to `[rw sync]` this form.

```
 19 nfs:
 20   enabled: true
 21   systemd name: nfs-server
 22   exports:
 23   - /home
 24   - /var/warewulf
 25   exports extended:
 26     - path: /opt
 27       options:
 28         - rw
 29         - sync
 30       mount: true
 ```
 
 ```
# This file was written by Warewulf (wwctl configure nfs)
/home 10.0.0.0/255.255.255.0(sync)
/var/warewulf 10.0.0.0/255.255.255.0(sync)
/opt 10.0.0.0/255.255.255.0([rw sync])
 ```

Reply:
We should be able to handle the default being enabled in the YAML unmarshal parsing. In theory, I think this should work, but I didn't test to confirm:

```
	Enabled     bool     `yaml:"enabled"  default:"true"`
```

Reply:
It looks like `gopkg.in/yaml.v2` doesn't support **default value**. --> https://github.com/go-yaml/yaml/issues/165
With this module, https://github.com/creasty/defaults , easy to support **default value**.

The following is git diff from this PR:

```
diff --git a/internal/pkg/warewulfconf/datastructure.go b/internal/pkg/warewulfconf/datastructure.go
index 620665f..0cfac05 100644
--- a/internal/pkg/warewulfconf/datastructure.go
+++ b/internal/pkg/warewulfconf/datastructure.go
@@ -3,6 +3,7 @@ package warewulfconf
 import (
        "github.com/hpcng/warewulf/internal/pkg/util"
        "github.com/hpcng/warewulf/internal/pkg/wwlog"
+       "github.com/creasty/defaults"
 )
 
 const ConfigFile = "/etc/warewulf/warewulf.conf"
@@ -43,12 +44,23 @@ type TftpConf struct {
 }
 
 type NfsConf struct {
-       Enabled     bool     `yaml:"enabled"`
+       Enabled     bool     `default: "true" yaml:"enabled"`
        Exports     []string `yaml:"exports"`
        ExportsExtended []*NfsExportConf `yaml:"exports extended"`
        SystemdName string   `yaml:"systemd name"`
 }
 
+func (s *NfsConf) UnmarshalYAML(unmarshal func(interface{}) error) error {
+    defaults.Set(s)
+
+    type plain NfsConf
+    if err := unmarshal((*plain)(s)); err != nil {
+        return err
+    }
+
+    return nil
+}
+
 type NfsExportConf struct {
        Path     string  `yaml:"path"`
        Options  string  `yaml:"options"`
```

and configuration file looks like this

```
nfs:
  enabled: true
  systemd name: nfs-server
  exports:
  - /var/warewulf
  exports extended:
    - path: /home
      options: rw,sync
      mount: true
    - path: /opt
      options: sync
      mount: false
```

and I set to nfs enabled to false, and execute `sudo wwctl configure nfs -d -v`,
then  I get following debug print produced by following code:

```
2021/11/08 06:41:41.604538 [ERROR] 	  NFS ENABLED OPTION: false
```

```
diff --git a/internal/app/wwctl/configure/nfs/main.go b/internal/app/wwctl/configure/nfs/main.go
index b9d5ec0..cb98917 100644
--- a/internal/app/wwctl/configure/nfs/main.go
+++ b/internal/app/wwctl/configure/nfs/main.go
@@ -47,7 +47,7 @@ func Configure(show bool) error {
                fmt.Fprintf(fstab, "tmpfs /run/shm tmpfs defaults 0 0\n")
                fmt.Fprintf(fstab, "sysfs /sys sysfs defaults 0 0\n")
                fmt.Fprintf(fstab, "proc /proc proc defaults 0 0\n")
-
+               wwlog.Printf(wwlog.ERROR, "NFS ENABLED OPTION: %t\n",controller.Nfs.Enabled)
                if controller.Nfs.Enabled {
                        exports, err := os.OpenFile("/etc/exports", os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
                        if err != nil {
```

Reply:
What is printed when you _remove_ the `enabled` key from the config file, i.e., did you verify that the default value is true when not explicitly set?

Reply:
@robgjansen
Yes, I verified.
If I remove `enabled` key from config file, this PR sets `true`(default value) to `enabled`.

```
nfs:
  systemd name: nfs-server
  exports:
  - /var/warewulf
```

Reply:
@panda1100 I've never used `github.com/creasty/defaults` before, but very cool implementation to set YAML/JSON defaults. But I also think it might be a tad bit overkill, as it would be pretty easy just to evaluate the enabled member. Right?